core/src/main/scala/com/criteo/cuttle/Executor.scala

```diff
@@ -81,23 +85,49 @@ case class Execution[S <: Scheduling](
   platforms: Seq[ExecutionPlatform],
   executionContext: ExecutionContext
 ) {
-  private[cuttle] val cancelSignal = Promise[Nothing]
-  def isCancelled = cancelSignal.isCompleted
-  val cancelled = cancelSignal.future
   private var waitingSeconds = 0
-  var startTime: Option[Instant] = None
-
-  def onCancelled(thunk: () => Unit) = cancelled.andThen {
-    case Failure(_) =>
+  private[cuttle] var startTime: Option[Instant] = None
+  private val cancelListeners = TSet.empty[CancellationListener]
+  private val cancelled = Ref(false)
+
+  def isCancelled = cancelled.single()
+  def onCancel(thunk: () => Unit): CancellationListener = {
+    val listener = new CancellationListener(this, thunk)
+    val alreadyCancelled = atomic { implicit txn =>
+      if(!cancelled()) {
+        cancelListeners += listener
+        false
+      }
+      else {
+        true
+      }
+    }
+    if(alreadyCancelled) {
       thunk()
-    case Success(_) =>
-      sys.error("Panic, the cancelled future can never succeed!")
+    }
+    listener
   }
-  def cancel()(implicit user: User): Boolean =
-    if (cancelSignal.tryFailure(ExecutionCancelledException)) {
+  private[cuttle] def removeCancelListener(listener: CancellationListener): Unit = atomic { implicit txn =>
+    cancelListeners -= listener
+  }
+
+  def cancel()(implicit user: User): Boolean = {
+    val (hasBeenCancelled, listeners) = atomic { implicit txn =>
+      if(cancelled()) {
+        (false, Nil)
+      } else {
+        cancelled() = true
+        val listeners = cancelListeners.snapshot
+        cancelListeners.clear()
+        (true, listeners)
+      }
+    }
+    if(hasBeenCancelled) {
       streams.debug(s"Execution has been cancelled by user ${user.userId}.")
-      true
-    } else false
+      listeners.foreach(_.thunk())
```

Comment:
Just curious if a thunk fails should the other be triggered ?

Reply:
Fixed